### PR TITLE
Upgrade and downgrade edition and datacache

### PR DESCRIPTION
**Pull request by algawivem**
Updates settings edition, allowing for in place upgrade to enterprise plus with data cache or downgrade to enterprise.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
sql: added support for in-place updates to the `edition` field in `google_sql_database_instance`
```

### CODE REVIEW

**Comment by algawivem** [comment]
@NickElliot could you take a look?

**Review by NickElliot — COMMENTED** [review]
Hi! Would it be possible to update the Upgrade case to also use the "testGoogleSqlDatabaseInstance_EditionConfig" and testGoogleSqlDatabaseInstance_EditionConfig_noEdition" tests in reverse order? Or does upgrading a MySQL db require the data cache enabling?

**Comment by algawivem** [comment]
Added `testGoogleSqlDatabaseInstance_EditionConfig` case. It is a Postgres config. Upgrade does not require data cache although we expect most customers to have it.

File: mmv1/third_party/terraform/services/sql/resource_sql_database_instance.go.erb [hunk 1]
```diff
@@ -1880,6 +1880,33 @@ func resourceSqlDatabaseInstanceUpdate(d *schema.ResourceData, meta interface{})
 		}
 	}
 
+	// Check if edition changed and patch with datacache.
+	if d.HasChange("settings.0.edition") {
```
Comment (NickElliot) [inline mmv1/third_party/terraform/services/sql/resource_sql_database_instance.go.erb:1884]: One question about functionality: is it intended that you can only update "tier" if edition is upgraded? With how this code is setup, the code would not process if only "tier" was updated.

**Review by NickElliot — COMMENTED** [review]
I have one question in a comment, but wanted to ask something broader as well. What is the functionality change in the request by sending the patch request with data cache as opposed to the current functionality for updating via an "update" command from the below lines for the "settings" block starting on line 1910? As it appears there is already support for "upgrading/downgrading" for these fields via that.

File: mmv1/third_party/terraform/services/sql/resource_sql_database_instance.go.erb (hunk 1 above)
Comment (algawivem) [reply mmv1/third_party/terraform/services/sql/resource_sql_database_instance.go.erb:1884]: Yes, this is intended; there are specific enterprise plus tiers that are only available if user specifies enterprise plus edition.

Comment (algawivem) [reply mmv1/third_party/terraform/services/sql/resource_sql_database_instance.go.erb:1884]: This is implementing: https://cloud.google.com/sql/docs/mysql/upgrade-cloud-sql-instance-to-enterprise-plus-in-place 
For in place upgrade, edition (along with tier and data cache) must be changed on its own. I can add a comment in the code.

Comment (NickElliot) [reply mmv1/third_party/terraform/services/sql/resource_sql_database_instance.go.erb:1885]: @algawivem Would this lead to a potential state of a user trying to upgrade "tier" with the state apply being successful, but then receiving a permadiff as it wouldn't properly update?

Comment (algawivem) [reply mmv1/third_party/terraform/services/sql/resource_sql_database_instance.go.erb:1885]: The argument "tier" is required so validate will fail if not provided.

If tier does not match the enterprise type, apply will fail for example: `Error: Error, failed to update instance settings for : googleapi: Error 400: Invalid request: Invalid Tier (db-perf-optimized-N-4) for (ENTERPRISE) Edition.`

**Review by NickElliot — APPROVED** [review]
Alright, LGTM then!

**Pull request merged**